// @nativescript/nx/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.typescriptVersion = exports.zonejsVersion = exports.rxjsVersion = exports.nsNgToolsVersion = exports.nsAngularVersion = exports.angularVersion = exports.nsAndroidRuntimeVersion = exports.nsIOSRuntimeVersion = exports.sassVersion = exports.nsWebpackVersion = exports.nsThemeVersion = exports.nsTypesVersion = exports.nsCoreVersion = exports.nsNxPluginVersion = exports.nxVersion = void 0;
exports.nxVersion = '^13.0.0';
exports.nsNxPluginVersion = '~2.0.0';
exports.nsCoreVersion = '~8.1.0';
exports.nsTypesVersion = '~8.1.0';
exports.nsThemeVersion = '~1.0.4';
exports.nsWebpackVersion = '~5.0.0';
exports.sassVersion = '^1.32.0';
exports.nsIOSRuntimeVersion = '~8.1.0';
exports.nsAndroidRuntimeVersion = '~8.1.1';
// Frontend frameworks
exports.angularVersion = '^12.0.0';
exports.nsAngularVersion = '^12.0.0';
exports.nsNgToolsVersion = '^12.0.0';
exports.rxjsVersion = '~6.6.0';
exports.zonejsVersion = '^0.11.4';
exports.typescriptVersion = '~4.3.5';
