// @nativescript/angular/esm2015/lib/cdk/frame-page/frame-page.component.js
import { Component, ElementRef, Input, Renderer2 } from '@angular/core';
import { Frame, Page } from '@nativescript/core';
import { frameMeta, registerElement } from '../../element-registry';
import { PageService } from './page.service';
export function customFrameComponentFactory(v) {
    return v.element.nativeElement;
}
export function customPageFactoryFromFrame(v) {
    return v.page;
}
registerElement('FramePage', () => Frame, {
    insertChild: (parent, child) => {
        if (child instanceof Page) {
            frameMeta.insertChild(parent, child);
            return;
        }
        parent['__ng_page__'].content = child;
    },
});
export class FramePageComponent {
    constructor(element, renderer) {
        this.element = element;
        this.actionBarHidden = false;
        this.page = renderer.createElement('Page');
        element.nativeElement['__ng_page__'] = this.page;
        renderer.appendChild(element.nativeElement, this.page);
    }
    ngOnChanges(changes) {
        if (changes.actionBarHidden && changes.actionBarHidden.previousValue !== changes.actionBarHidden.currentValue) {
            if (this.page) {
                this.page.actionBarHidden = !!this.actionBarHidden;
            }
        }
    }
    ngAfterViewInit() {
        if (this.page) {
            this.page.actionBarHidden = !!this.actionBarHidden;
        }
    }
}
FramePageComponent.decorators = [
    { type: Component, args: [{
                // eslint-disable-next-line @angular-eslint/component-selector
                selector: 'FramePage',
                template: `<ng-content></ng-content>`,
                providers: [
                    {
                        provide: Frame,
                        useFactory: customFrameComponentFactory,
                        deps: [FramePageComponent],
                    },
                    {
                        provide: Page,
                        useFactory: customPageFactoryFromFrame,
                        deps: [FramePageComponent],
                    },
                    PageService,
                ]
            },] }
];
FramePageComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
FramePageComponent.propDecorators = {
    actionBarHidden: [{ type: Input }]
};
