// @nativescript/angular/esm2015/lib/views/utils.js
import { ContentView, LayoutBase, ProxyViewContainer, View } from '@nativescript/core';
import { InvisibleNode } from './invisible-nodes';
export function isDetachedElement(element) {
    return element && element.meta && element.meta.skipAddToDom;
}
export function isView(view) {
    return view instanceof View;
}
export function isInvisibleNode(view) {
    return view instanceof InvisibleNode;
}
export function isLayout(view) {
    return view instanceof LayoutBase;
}
export function isContentView(view) {
    return view instanceof ContentView;
}
export function getFirstNativeLikeView(view, extractFromNSParent = false) {
    if (view instanceof ProxyViewContainer) {
        if (view.getChildrenCount() === 0) {
            return null;
        }
        return getFirstNativeLikeView(view.getChildAt(0));
    }
    if (extractFromNSParent) {
        // const node = view.parentNode;
        detachViewFromParent(view);
        // view.parentNode = node;
    }
    return view;
}
export function detachViewFromParent(view) {
    const parent = view === null || view === void 0 ? void 0 : view.parent;
    if (!parent) {
        return;
    }
    if (parent.meta && parent.meta.removeChild) {
        parent.meta.removeChild(parent, view);
    }
    else if (isLayout(parent)) {
        parent.removeChild(view);
    }
    else if (isContentView(parent) && parent.content === view) {
        parent.content = null;
    }
    else if (isView(parent)) {
        parent._removeView(view);
    }
}
