// @nativescript/angular/esm2015/lib/loading.service.js
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class NativeScriptLoadingService {
    constructor() {
        this.mainModuleReady$ = new BehaviorSubject(false);
        this.readyToDestroy$ = new BehaviorSubject(true);
        this.onMainModuleReady$ = this.mainModuleReady$.pipe(filter((ready) => ready));
    }
    /**
     * delays destroying this module until `notifyReadyToDestroy()`.
     * remember to call `notifyReadyToDestroy()` when done!
     */
    waitUntilNotified() {
        this.readyToDestroy$.next(false);
    }
    /**
     * notifies this module is ready to be destroyed
     */
    notifyReadyToDestroy() {
        this.readyToDestroy$.next(true);
    }
    isMainModuleReady() {
        return this.mainModuleReady$.value;
    }
    ngOnDestroy() {
        this.readyToDestroy$.complete();
        this.mainModuleReady$.complete();
    }
    /**
     * This funcion is called by the bootstrap code when the app is ready
     * @internal
     */
    _notifyMainModuleReady() {
        this.mainModuleReady$.next(true);
    }
}
NativeScriptLoadingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NativeScriptLoadingService_Factory() { return new NativeScriptLoadingService(); }, token: NativeScriptLoadingService, providedIn: "root" });
NativeScriptLoadingService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
