// @nativescript/angular/esm2015/lib/platform-nativescript.js
import { Sanitizer, createPlatformFactory, platformCore } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { NativeScriptPlatformRefProxy } from './platform-ref';
import { AppHostView } from './app-host-view';
import { Color, GridLayout } from '@nativescript/core';
import { defaultPageFactory, PAGE_FACTORY } from './tokens';
const ɵ0 = defaultPageFactory;
export const defaultPageFactoryProvider = { provide: PAGE_FACTORY, useValue: ɵ0 };
export class NativeScriptSanitizer extends Sanitizer {
    sanitize(_context, value) {
        return value;
    }
}
// TODO: when angular finally exports their scheduler token for ivy CD, provide our own with queueMacroTask
// export function schedulerFactory() {
//   return (fn: any) => {
//     console.log('SCHEDULER');
//     setTimeout(fn, 0);
//   };
// }
export class NativeScriptDocument {
    constructor() {
        // Required by the AnimationDriver
        this.body = {
            isOverride: true,
        };
    }
    createElement(tag) {
        throw new Error('NativeScriptDocument is not DOM Document. There is no createElement() method.');
    }
}
export const COMMON_PROVIDERS = [defaultPageFactoryProvider, { provide: Sanitizer, useClass: NativeScriptSanitizer, deps: [] }, { provide: DOCUMENT, useClass: NativeScriptDocument, deps: [] }];
export const platformNativeScript = createPlatformFactory(platformCore, 'nativescriptDynamic', COMMON_PROVIDERS);
/**
 * @deprecated use runNativeScriptAngularApp instead
 */
export const platformNativeScriptDynamic = function (options, extraProviders) {
    console.log('platformNativeScriptDynamic is deprecated, use runNativeScriptAngularApp instead');
    options = options || {};
    extraProviders = extraProviders || [];
    const ngRootView = new AppHostView(new Color(options.backgroundColor || 'white'));
    let launchView = options.launchView;
    if (!launchView && options.async) {
        launchView = new GridLayout();
        launchView.backgroundColor = options.backgroundColor || 'white';
    }
    return new NativeScriptPlatformRefProxy(platformNativeScript([...extraProviders]), launchView);
};
export { ɵ0 };
